### ISSUE TITLE
Memcache serialization

### ISSUE THREAD

**Issue by HankBrown** [yiisoft/yii2#14640]
### What steps will reproduce the problem?

1. Put this key into memcache:
```

$ telnet 192.168.0.7 11211
Trying 192.168.0.7...
Connected to 192.168.0.7.
Escape character is '^]'.
get MYCACHEID
VALUE MYCACHEID 0 67
a:2:{s:10:"time_stamp";i:1502667167;s:4:"data";s:11:"MyTestValue";}
END

```
In my case, another (non-Yii2) app has done so.

2. Try to read it...

```
    'components' => [
...
        'mcache' => [                                   // *** NB: mcache not cache
            'class' => 'yii\caching\MemCache',
            'servers' => [
                [
                    //'host' => '89.185.228.213', // dex213
                    'host' => '192.168.0.7',    // dex213.intfrom 321
                    'port' => 11211,
                ],
            ],
        ],

```

```
        $mc = \Yii::$app->mcache;   // mem cached
        $mc->useMemcached = true;
        $result=$mc->get('MYCACHEID'),

```        

### What is the expected result?

```
array
(
    [time_stamp] => 1502667167
    [data] => 'MyTestValue'
)
```
### What do you get instead?

Undefined offset 1 in  in /home/brownhanky/sites/www/vendor/yiisoft/yii2/caching/Cache.php

I can see this is a problem:
```

        } elseif ($this->serializer === null) {
 
Dump::d('NULL SERIALIZER CACHE GET KEY',$key,'VALUE',$value);
            $value = unserialize($value);
        } else {
            $value = call_user_func($this->serializer[1], $value);
        }
Dump::d('AFTER SERIALIZER CACHE GET KEY',$key,'VALUE',$value);
        if (is_array($value) && !($value[1] instanceof Dependency && $value[1]->isChanged($this))) {
            return $value[0];

The contents of my Dump is:

[2017-08-13 04:42:26]
'NULL SERIALIZER CACHE GET KEY'
'MYCACHEID'
'VALUE'
'a:2:{s:10:"time_stamp";i:1502667167;s:4:"data";s:11:"MyTestValue";}'
[2017-08-13 04:42:26]
'AFTER SERIALIZER CACHE GET KEY'
'MYCACHEID'
'VALUE'
array
(
    [time_stamp] => 1502667167
    [data] => 'MyTestValue'
) 

```

Ok. Yii2 is looking for an array with integer indices? 

### Additional info

| Q                | A
| ---------------- | ---
| Yii version      | 2.0.? composer:  "yiisoft/yii2": "~2",
| PHP version      | 5.5.9-1ubuntu4.4
| Operating system | Linux dex321 3.13.0-35-generic #62-Ubuntu SMP Fri Aug 15 01:58:42 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

http://www.yiiframework.com/doc-2.0/yii-caching-cache.html#$serializer-detail

"The first element specifies the serialization function, and the second the deserialization function."
How? My guesses fail.

"Also, some implementations of the cache can not correctly save and retrieve data different from a string type."
Maybe that is my case. But how would I know?!

I do not really think this is a bug, Yii2 has not steered me wrong yet.
But the docs and/or my frontal cortex do not have enough information.
I think...

**Comment by samdark** [comment]
Would you please post full error message including line numbers and stack trace?

**Comment by HankBrown** [comment]
```
		2017-08-14 06:31:08 [75.157.241.9][-][-][error][yii\base\ErrorException:8] exception 'yii\base\ErrorException' with message 'Undefined offset: 1' in /home/mysite/sites/www/vendor/yiisoft/yii2/caching/Cache.php:122
		Stack trace:
		#0 /home/mysite/sites/www/vendor/yiisoft/yii2/caching/Cache.php(122): yii\base\ErrorHandler->handleError(8, 'Undefined offse...', '/home/brownhank...', 122, Array)
		#1 /home/mysite/sites/www/frontend/controllers/TestController.php(688): yii\caching\Cache->get('MYCACHEID')
		#2 [internal function]: frontend\controllers\TestController->actionWtf()
		#3 /home/mysite/sites/www/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
		#4 /home/mysite/sites/www/vendor/yiisoft/yii2/base/Controller.php(156): yii\base\InlineAction->runWithParams(Array)
		#5 /home/mysite/sites/www/vendor/yiisoft/yii2/base/Module.php(523): yii\base\Controller->runAction('wtf', Array)
		#6 /home/mysite/sites/www/vendor/yiisoft/yii2/web/Application.php(102): yii\base\Module->runAction('test/wtf', Array)
		#7 /home/mysite/sites/www/vendor/yiisoft/yii2/base/Application.php(380): yii\web\Application->handleRequest(Object(yii\web\Request))
		#8 /home/mysite/sites/www/frontend/web/index.php(90): yii\base\Application->run()
		#9 {main}
		2017-08-14 06:31:08 [75.157.241.9][-][-][info][application] $_GET = [
		    'r' => 'test/wtf'
		]

```
Not sure how to find exact  version of Yii2. I see below and now worry about a dev version.
Appreciate any advice thank you.
vendor/yiisoft/yii2/composer.json : 

```
    "extra": {
        "branch-alias": {
            "dev-master": "2.0.x-dev"
        }
    }

```

**Comment by samdark** [comment]
It's basically not what Yii expects to be put into cache. If third party app wants to write there and expect it read properly, format should be what Yii expects:

[value, dependency]

**Comment by klimov-paul** [comment]
You should be able to use your code in case you disabled `$serializer` at cache:

```php
return [
    'components' => [
        'cache' => [
            'class' => 'yii\caching\MemCache',
            'serializer' => false,
        ],
    ],
];
```

However you'll have to unserialiize the value on your own.

**Comment by HankBrown** [comment]
@samdark : Thank you for a prompt response.
@klimov-paul: Perfect.

I suggest the confusion is over the very name "memcache".
I can be used for a sophisticated caching mechanism, or a simple shared/fast key/value store.
http://www.yiiframework.com/doc-2.0/yii-caching-memcache.html
"MemCache implements a cache application component based on memcache and memcached."
**(so if you just want to use a key/value store, build your own class or set serializer false).**

Sorry no PR for docs.

**Comment by klimov-paul** [comment]
Will be solved by #14622
